crates/env/src/engine/on_chain/impls.rs

```diff
@@ -499,21 +499,30 @@ impl TypedEnvBackend for EnvInstance {
         let out_address = &mut scoped.take(1024);
         let salt = params.salt_bytes().as_ref();
         let out_return_value = &mut scoped.take_rest();
-        // We currently do nothing with the `out_return_value` buffer.
-        // This should change in the future but for that we need to add support
-        // for constructors that may return values.
-        // This is useful to support fallible constructors for example.
-        ext::instantiate(
+
+        let instantiate_result = ext::instantiate(
             enc_code_hash,
             gas_limit,
             enc_endowment,
             enc_input,
             out_address,
             out_return_value,
             salt,
-        )?;
-        let account_id = scale::Decode::decode(&mut &out_address[..])?;
-        Ok(account_id)
+        );
+
+        match instantiate_result {
+            Ok(()) => {
+                let account_id = scale::Decode::decode(&mut &out_address[..])?;
+                Ok(Ok(account_id))
+            }
+            Err(ext::Error::CalleeReverted) => {
+                // We don't wrap manually with an extra `Err` like we do in the `Ok` case since the
+                // buffer already comes back in the form of `Err(LangError)`
+                let out = scale::Decode::decode(&mut &out_return_value[..])?;
```

Comment:
Couldn't there exist a contract that would revert the contract but return `Ok(())` e.g.
```
::ink::env::return_value::<::ink::ConstructorResult<()>>(
    ::ink::env::ReturnFlags::new_with_reverted(true),
    Ok(()), 
);
```
This could mislead the caller into handling the `Ok` case?

Reply:
Hmm yeah this is a good observation. I've added a test that shows that you can indeed revert from a constructor and continue execution by returning `Ok(AccountId)`. I've also added a quick patch which asserts that the content in the buffer is an error (the test passes without that patch).

I'm not entirely sure what the implications of either case are though...
- In the `Revert + Ok` case, misleading the caller seems like the main reason to do this, but maybe there are legit reasons to do encode an `Ok` back? I can't think of any though.
- In the `Revert + Err` case, the callee can encode the "wrong" `LangError` into the output buffer, potentially misleading the caller too, and there's nothing we can do to help there.

Either way it'll be up to the contract developer to make sure they understand that there are risks to calling other contracts.

Reply:
@athei would like to hear your opinion here

Reply:
The assumption is that the user is not calling `return_value` directly, and if they are they are using it in an ABI compatible way which is either `Revert + 'Err(LangError)` or `Revert + Ok(Err(ContracErr))`.

So perhaps we could return an `Err(LangError)` here if this ABI is not satisfied, e.g. as mentioned in https://github.com/paritytech/ink/pull/1512#discussion_r1036968362

Reply:
Okay maybe what we want is something along the lines of
```
let out = ink_primitives::ConstructorResult::<()>::decode(
    &mut &out_return_value[..],
)?;
match out {
    // This branch when a fallible constructor returns `Err`
    Ok(()) => Err(ext::Error::CalleeReverted),
    /// This branch handles an actual `LangErr`
    Err(lang_err) => Ok(Err(lang_err))
}
```